docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+topic.
+
+These concepts are not visible to applications.  They are optionally visible
+to system administrators optionally:
+- system monitoring telemetry, where we would surface indicators for how many partitions
+  and topics are currently hibernated.
+- tunable configuration properties, where they could control the delay til hibernation
+  if needed.
+
+
+# Reference-level explanation
+
+## High level design
+
+The topic table, which stores the definition of the topics & partitions
+on the system, is driven by controller log commands.  New commands
+are added for hibernate and wake.  Nodes in a partition's raft group
+notice the hibernation of the topic, and shut down the raft group.
+
+There is a "hibernating" intermediate state where members of the
+partitions' raft groups are responsible for refusing writes, and
+reporting the highest offset back to the controller so that it can
+reliably store the frozen offset state of the topic when storing
+the controller message that finalizes the hibernation.
+
+Once the topic is hibernated, all its partitions will report a
+fictional leader, so that clients know where to send metadata requests,
+and so that monitoring systems do not incorrectly flag the partition
+as being in a problematic state.
+
+On startup, a hibernated partition's storage resources are not
+recovered or replayed.  Hibernated partitions are not subject to
+continuous retention policies, BUT the controller can notice if
+a partition has been hibernated for longer than its retention.ms
+limit, and wake it up for long enough to do housekeeping.
+
+We should still be able to serve metadata requests for idle topics,
+because otherwise any monitoring system will end up re-awakening
+the topics whenever querying them.
+
+## How do we decide when to hibernate?
+
+We don't persist an `atime` type stat for partitions, and rightly so,
+because otherwise we'd be writing metadata every time someone did
+a read.
+
+Detecting an idle partition at runtime therefore relies on a partition
+staying up with stable leadership long enough to reach the idle timeout.
+That's fine if the timeout is a few hours long (if we're changing leaders
+more often than that then we've got bigger problems) but more problematic
+if we have a long idle timeout (perhaps a day or more) with a reasonable
+chance of leadership changes or node restarts before we reach it.
+
+If we wanted to support very long idle timeouts better, we could
+do so by adding something like a kvstore record of the access time
+that is updated at most every N minutes, and when a leader is elected
+it primes its access time from that if it is the same node that was
+also the previous leader.
+
+## What resources are we saving by hibernating?
+
+- Avoid creating new files on log roll during startup
+- Avoid replaying segments during startup
+- No memory overhead from storing segment indices, list of segments, etc.
+- No raft heartbeats for hibernated topics
```

Comment:
For this one, I had an idea:
- Replace empty "timer" heartbeats in per-NTP groups with an abort source that is triggered when cluster-level quorum changes (a node goes down, cluster startup, shutdown, add, node, topic/partition map changes etc.). The main cluster group (i.e. controller) essentially provides heartbeat services by proxy for the purposes of detecting group  changes, etc.

Reply:
Currently, nodes can change leadership for reasons other than a node going entirely down.  e.g. a node that stalls for long enough will miss heartbeats and followers will start an election.  Also partition movement, leadership transfer, etc.  It's also important that the heartbeats are traveling across the same network routes that would be used for data, so that a heartbeat is proof that replication would also work.

Not that there aren't ways to optimize this for the steady state, but it's probably quite a significant change in its own right.

Reply:
> Currently, nodes can change leadership for reasons other than a node going entirely down. e.g. a node that stalls for long enough will miss heartbeats and followers will start an election.
> Also partition movement, leadership transfer, etc. It's also important that the heartbeats are traveling across the same network routes that would be used for data, so that a heartbeat is proof that replication would also work.
> 

Yeah, thought about this some. Two causes of group membership change: failures (which HB's detect) and planned changes.

**Failures**: Not sure we want to live in a world where we handle partial node failure; complex and split-brain-prone. The optimal failure units IMO are things like link, node, disk. We can optimize scaling if our heartbeats and error handling have "mechanical sympathy" or map closely to this reality. This is the argument for link-level heartbeats.
With our "mostly independent seastar shards" design though; there is the question of; do we need to monitor per-connection or per link. Per-link can be significantly more efficient, and one could argue that if you have connections / fibers / etc. getting completely starved, you have other problems you should fix instead of doing more work to accommodate it?

**Planned changes**: Not triggered by HB timeout but executed. In this case the "abort source" that needs to kick/restart 
affected fibers/operations can be driven similarly, but instead of changing the top-level quorum, you can be more targeted, .i.e. possible to only restart operations whose NTP overlaps with the delta of the topic/partition table. There are simple solutions or more complex ones that scale better there.

> Not that there aren't ways to optimize this for the steady state, but it's probably quite a significant change in its own right.

For sure.. I'm thinking ahead a bit. These simplifying assumptions can also help out a lot with all the consistency cases that happen (think jepsen scenarios, narrowing possibilities for split brains). And factoring out the "do this operation with quorum state and retries" part can make writing and understanding the code a lot easier too.